File: 303_Binary_Sum_Query_Immutable/sumRange.py
class NumArray(object):
    def __init__(self, nums):
        """
        initialize your data structure here.
        :type nums: List[int]
        """
        # sums[i]: sum of 0th to (i - 1)th elements 
        n_length = len(nums) + 1
        self.sums = [0] * n_length
        for i in xrange(1, n_length):
            self.sums[i] = self.sums[i - 1] + nums[i - 1]
        print self.sums
            
    def sumRange(self, i, j):
        """
        sum of elements nums[i..j], inclusive.
        :type i: int
        :type j: int
        :rtype: int
        """
        return self.sums[j + 1] - self.sums[i]

        

nums = [1, 2, 3]
print nums
numArray = NumArray(nums) 
print numArray.sumRange(0 , 1)

File: 303_Binary_Sum_Query_Immutable/sumRange.cpp
class NumArray {
public:
    NumArray(vector<int> &nums) {
        int n = nums.size();
        sums.resize(n + 1);
        sums[0] = 0;
        for(int i = 1; i <= n; i++) {
            sums[i] = sums[i - 1] + nums[i - 1];
        }
    }

    int sumRange(int i, int j) {
        return (sums[j + 1] - sums[i]);
    }

private:
    // sums[i] = sum(nums[0]...nums[i-1])
    vector<int> sums;
};
